### DISCUSSION TITLE
Add a duckdb_attributes() function to help reverse engineer STRUCT and other types

### DISCUSSION THREAD

**Discussion by lukaseder** [duckdb/duckdb#8832] [Ideas]
When creating a type like this:

```sql
create type u as struct (i int, j int);
```

It's currently not possible (I think?) to reverse engineer it from the existing `duckdb_xyz()` table functions. I can find the type like this:

```sql
select *
from duckdb_types()
where type_name = 'u';
```

But that just contains the following information:

```
|database_name|database_oid|schema_name|schema_oid|type_oid|type_name|type_size|logical_type|type_category|internal|labels|
|-------------|------------|-----------|----------|--------|---------|---------|------------|-------------|--------|------|
|duckdb       |916         |main       |918       |2,129   |u        |0        |STRUCT      |COMPOSITE    |false   |      |
```

The `STRUCT` is "empty", and I don't see any other table that would contain more information.

A workaround is to create a dummy table or view containing the type:

```sql
select data_type
from duckdb_columns
where table_name = 'dummy';
```

This now produces a string form of the `STRUCT` type:

```
|data_type                   |
|----------------------------|
|STRUCT(i INTEGER, j INTEGER)|
```

~I guess I can work with that~ (turns out, I can't, there is no formal link to the `STRUCT` type: https://github.com/duckdb/duckdb/issues/8838), but it would be nice to have more formal information. The use-case is source code generation in jOOQ.

**Comment by Tishj** [comment]
I think it's advised to use `DESCRIBE` in this case
example using the python binding:
```py
>>> duckdb.sql("describe select * from tbl")
┌─────────────┬──────────────────────────────┬─────────┬─────────┬─────────┬─────────┐
│ column_name │         column_type          │  null   │   key   │ default │  extra  │
│   varchar   │           varchar            │ varchar │ varchar │ varchar │ varchar │
├─────────────┼──────────────────────────────┼─────────┼─────────┼─────────┼─────────┤
│ a           │ STRUCT(a INTEGER, b BOOLEAN) │ YES     │ NULL    │ NULL    │ NULL    │
└─────────────┴──────────────────────────────┴─────────┴─────────┴─────────┴─────────┘
```

**Reply by Tishj** [reply]
Also if you're using the python bindings, there is support for dissecting the nested type
```
>>> duckdb.table("tbl").types
[STRUCT(a INTEGER, b BOOLEAN)]
>>> duckdb.table("tbl").types[0]
STRUCT(a INTEGER, b BOOLEAN)
>>> duckdb.table("tbl").types[0].children
[('a', INTEGER), ('b', BOOLEAN)]
```

**Reply by lukaseder** [reply]
I'm using JDBC

**Reply by Tishj** [reply]
Hmm I'm not aware of any JDBC specific methods for this

**Reply by lukaseder** [reply]
I'm not requesting any JDBC specific methods. I'm requesting SQL functionality, i.e. a new table/function.

**Reply by Tishj** [reply]
I'm aware, I'm just looking for work arounds.
Feel free to make a suggestion, the problem is clear, the solution not so much

**Reply by lukaseder** [reply]
> I'm aware, I'm just looking for work arounds.

Ah, I see, thanks :). Well, my workaround suffices for my own needs. I have a SQL parser that will be capable of traversing any nested `STRUCT` specification, so that'll do for now.

> Feel free to make a suggestion, the problem is clear, the solution not so much

The standard SQL `INFORMATION_SCHEMA.ATTRIBUTES` contains the UDT ("STRUCT") attribute information. PostgreSQL also stores them in `pg_attribute`. Other RDBMS have their own dictionary views, e.g. Oracle uses `ALL_TYPE_ATTRS`.

Here's a PostgreSQL example:

```sql
create type u as (i int, j int);

select *
from information_schema.attributes
where udt_name = 'u'
order by ordinal_position;
```

Result:

```
|udt_catalog|udt_schema|udt_name|attribute_name|ordinal_position|attribute_default|is_nullable|data_type|character_maximum_length|character_octet_length|character_set_catalog|character_set_schema|character_set_name|collation_catalog|collation_schema|collation_name|numeric_precision|numeric_precision_radix|numeric_scale|datetime_precision|interval_type|interval_precision|attribute_udt_catalog|attribute_udt_schema|attribute_udt_name|scope_catalog|scope_schema|scope_name|maximum_cardinality|dtd_identifier|is_derived_reference_attribute|
|-----------|----------|--------|--------------|----------------|-----------------|-----------|---------|------------------------|----------------------|---------------------|--------------------|------------------|-----------------|----------------|--------------|-----------------|-----------------------|-------------|------------------|-------------|------------------|---------------------|--------------------|------------------|-------------|------------|----------|-------------------|--------------|------------------------------|
|postgres   |public    |u       |i             |1               |                 |YES        |integer  |                        |                      |                     |                    |                  |                 |                |              |32               |2                      |0            |                  |             |                  |postgres             |pg_catalog          |int4              |             |            |          |                   |1             |NO                            |
|postgres   |public    |u       |j             |2               |                 |YES        |integer  |                        |                      |                     |                    |                  |                 |                |              |32               |2                      |0            |                  |             |                  |postgres             |pg_catalog          |int4              |             |            |          |                   |2             |NO                            |
```

**Reply by Tishj** [reply]
Hmm I see, that doesn't directly translate to duckdb, because `create type as ...` is pretty much just syntactic sugar
It's perfectly valid to write
```sql
create table tbl (
	i struct(
		a map(varchar, bool[]),
		b struct(
			i bool,
			j int[]
		)
	),
	j struct(
		a struct(
			a int
		)
	)[][][]
);
```

The method you proposed doesn't scale for this
Only when every "complex" type is a separate entry

**Reply by Tishj** [reply]
I agree that it should be possible to query, I'm just not sure yet how this should be done

**Reply by lukaseder** [reply]
I can see two options:

- Create nominal types with a system name for every structural type (or even without a name, since you have surrogate keys in `DUCKDB_TYPES.TYPE_OID`, the name can be optional!)
- Keep structural types around

As far as precedents in other RDBMS are concerned, the standard SQL `INFORMATION_SCHEMA` has an `ELEMENT_TYPES` view that allows for describing arbitrary levels of arrays. I'm not aware of any way of declaring composite types ("UDTs") in a structural way, so the SQL standard won't help here. The standard uses identifiers as natural keys for schema objects, though most implementations have some additional surrogate key as well.

**Reply by lukaseder** [reply]
Just to be sure: While I think in the long run, a flat SQL representation like other RDBMS have would be desirable, in the short term, being able to access *any* type information at all is important for third party tools.
